Xarray GPU optimization

negin513 opened the pull request:
Contributors: @negin513, @weiji14 , @TomAugspurger , @maxrjoes, @akshaysubr, @kafitzgerald

TomAugspurger commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,223 @@
+
+## Introduction
+
+In large-scale geospatial AI and machine learning workflows, data loading is often the main bottleneck. Traditional pipelines rely on CPUs to transfer massive datasets from disk to GPU memory, consuming resources and limiting scalability and effective use of GPU resources.
```
```suggestion
In large-scale geospatial AI and machine learning workflows, data loading is often the main bottleneck. Traditional pipelines rely on CPUs to preprocess and transfer massive datasets from storage to GPU memory, consuming resources and limiting scalability and effective use of GPU resources.
```

TomAugspurger commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,223 @@
+
+In large-scale geospatial AI and machine learning workflows, data loading is often the main bottleneck. Traditional pipelines rely on CPUs to transfer massive datasets from disk to GPU memory, consuming resources and limiting scalability and effective use of GPU resources.
+
+To tackle this issue, a team from the [National Center for Atmospheric Research (NSF-NCAR)](https://ncar.ucar.edu) and [Development Seed](https://developmentseed.org) with mentors from [NVIDIA](https://www.nvidia.com) participated in a GPU hackathon to demonstrate how AI/ML workflows in Earth system sciences can benefit from GPU-native workflows using tools such as Zarr, kvikIO, and DALI.
```
```suggestion
To tackle this issue, a team from the [National Center for Atmospheric Research (NSF-NCAR)](https://ncar.ucar.edu) and [Development Seed](https://developmentseed.org) with mentors from [NVIDIA](https://www.nvidia.com) participated in a GPU hackathon to demonstrate how AI/ML workflows in Earth system sciences can benefit from GPU-native workflows using tools such as [Zarr](https://zarr.readthedocs.io/), [KvikIO](https://docs.rapids.ai/api/kvikio/stable/), and [DALI](https://developer.nvidia.com/dali).
```

TomAugspurger commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,223 @@
+- Transforming / preprocessing data (often CPU-bound).
+- Feeding the data into GPUs for training or inference.
+
+Although GPU compute is incredibly fast, the CPU can become a bottleneck when dealing with large dataset.
```
```suggestion
Although GPU compute is incredibly fast, the CPU can become a bottleneck when dealing with large datasets.
```

TomAugspurger commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,223 @@
+
+### Data & Code Overview
+
+For this hackathon, we developed a benchmark of training a U-NET (with ResNet backend) training on ERA-5 Dataset to predict next time steps. The U-Net model is implemented in PyTorch and the training pipeline is built using PyTorch DataLoader. The model can be trained on a single GPU or multiple GPUs using Distributed Data Parallel (DDP) for parallelization.
```
```suggestion
For this hackathon, we developed a benchmark of training a U-NET (with ResNet backend) model on the ERA-5 Dataset to predict next time steps. The U-Net model is implemented in PyTorch and the training pipeline is built using PyTorch DataLoader. The model can be trained on a single GPU or multiple GPUs using Distributed Data Parallel (DDP) for parallelization.
```

TomAugspurger reviewed (commented):
Thanks for writing this up!

dcherian commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,223 @@
+
+In large-scale geospatial AI and machine learning workflows, data loading is often the main bottleneck. Traditional pipelines rely on CPUs to preprocess and transfer massive datasets from storage to GPU memory, consuming resources and limiting scalability and effective use of GPU resources.
+
+To tackle this issue, a team from the [National Center for Atmospheric Research (NSF-NCAR)](https://ncar.ucar.edu) and [Development Seed](https://developmentseed.org) with mentors from [NVIDIA](https://www.nvidia.com) participated in a GPU hackathon to demonstrate how AI/ML workflows in Earth system sciences can benefit from GPU-native workflows using tools such as [Zarr](https://zarr.readthedocs ...
```
Some institution logos here could be nice visually

negin513 replied:
Good Idea! Added!

negin513 replied:
BTW, you meant the logo of NCAR, devseed, etc? That is added now at the end of the post, but I am open to bring them higher. 

I can also include the logos of packages( i.e. cupy, etc.) but not sure if that was what you meant.

dcherian commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,223 @@
+---
+
+# Accelerating AI/ML Workflows in Earth Sciences with GPU-Native Xarray and Zarr (and more!)
+
```
Let's add a TL; DR

dcherian commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,223 @@
+  - name: Katelyn Fitzgerald
+    github: kafitzgerald
+
+summary: 'How to accelerate AI/ML workflows in Earth Sciences with GPU-native Xarray and Zarr.'
```
Can we make this more direct? "X% speedup" or "XMBps throughput"?

negin513 replied:
Added ~17X all in all improvement!💪💪💪

dcherian commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,223 @@
+
+### Data & Code Overview
+
+For this hackathon, we developed a benchmark of training a U-NET (with ResNet backend) model on the ERA-5 Dataset to predict next time steps. The U-Net model is implemented in PyTorch and the training pipeline is built using PyTorch DataLoader. The model can be trained on a single GPU or multiple GPUs using Distributed Data Parallel (DDP) for parallelization.
```
Let's link to the Repo here before referring to files

dcherian commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,223 @@
+
+For this hackathon, we developed a benchmark of training a U-NET (with ResNet backend) model on the ERA-5 Dataset to predict next time steps. The U-Net model is implemented in PyTorch and the training pipeline is built using PyTorch DataLoader. The model can be trained on a single GPU or multiple GPUs using Distributed Data Parallel (DDP) for parallelization.
+
+-- TODO : Add an example image.
```
flagging this TODO

dcherian commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,223 @@
+- Transforming / preprocessing data (often CPU-bound).
+- Feeding the data into GPUs for training or inference.
+
+Although GPU compute is incredibly fast, the CPU can become a bottleneck when dealing with large datasets.
```
```suggestion
Although GPU compute is incredibly fast, the CPU can become a bottleneck when dealing with large datasets (profiles discussed below).
```

dcherian commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,223 @@
+
+To tackle this issue, a team from the [National Center for Atmospheric Research (NSF-NCAR)](https://ncar.ucar.edu) and [Development Seed](https://developmentseed.org) with mentors from [NVIDIA](https://www.nvidia.com) participated in a GPU hackathon to demonstrate how AI/ML workflows in Earth system sciences can benefit from GPU-native workflows using tools such as [Zarr](https://zarr.readthedocs ...
+
+In this post, we share our hackathon experience, the integration strategies we explored, and the performance gains we achieved to highlight how modern tools can transform data-intensive workflows.
```
Let's call out the hackathon and its funders by name (useful in this _climate_ hehe)

dcherian replied:
I see this at the end, but it would be nice to call it out up top

negin513 replied:
Done. Thanks for the suggestion. 🥇

negin513 replied:
100% agreed. Linked them at the beginning of the blog post and would be happy to highlight them more too. :-)

dcherian commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,223 @@
+During the hackathon, we tested the following strategies to improve the data loading performance:
+
+1. Optimized Chunking & Compression
+   - We explored different chunking and compression strategies to optimize the data loading performance. We found that using Zarr v3 with optimized chunking and compression significantly improved the data loading performance.
```
I would either delete this summary or shorten and add a summary to the remaining items.

Some links to nvcomp and DALI would be great too

dcherian commented on src/posts/gpu-pipeline/index.md:
````diff
@@ -0,0 +1,223 @@
+
+[ TODO: add a figure showing this -- technically decompression is still done on CPU. ]
+
+(TODO ongoing work) Eventually with this [cupy-xarray Pull Request merged](https://github.com/xarray-contrib/cupy-xarray/pull/70) (based on earlier work at https://xarray.dev/blog/xarray-kvikio), this can be simplified to:
+
+```python
+import cupy_xarray
+
+ds = xr.open_dataset(filename_or_obj="/tmp/air-temp.zarr", engine="kvikio")
+assert isinstance(ds.air.data, cp.ndarray)
+```
````
This could go in a future work section at the end

weiji14 replied:
Yeah, I'm not sure if this API is feasible or even desirable (have tried to implement this in https://github.com/xarray-contrib/cupy-xarray/pull/70, but no luck yet patching the buffer protocol). So ok to move this towards the end.

dcherian commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,223 @@
+
+![image](https://hackmd.io/_uploads/H1YVp6tR1l.png)
+
+(TODO insert better nsight profiling figure than above showing overlapping CPU and GPU compute)
```
that would be really nice!

dcherian commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,223 @@
+- Consider using GPU Direct Storage (GDS) for optimal performance, but be aware of the setup and configuration required.
+- GPU Direct Storage (GDS) can be an improvement for data-intensive workflows, but requires some setup and configuration.
+- NVIDIA DALI is a powerful tool for optimizing data loading, but requires some effort to integrate into existing workflows.
+- GPU-based decompression is a promising area for future work, but requires further development and testing.
```
Icechunk!

dcherian commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,223 @@
+---
+title: 'Accelerating AI/ML Workflows in Earth Sciences with GPU-Native Xarray and Zarr (and more!)'
```
```suggestion
title: 'GPU-Native Earth Science AI/ML Workflows Xarray, Zarr, DALI, and nvcomp'
```

weiji14 commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,223 @@
+For more optimal performance, consider:
+
+1. Storing the data without compression (if not transferring over a network), as decompressing data can slow down read speeds. But see also GPU decompression with nvCOMP below. :wink:
+2. Concatenating several data variables together **if** a single chunk size is too small (<1MB), at the expense of reducing readability of the Zarr store. Having too many small chunks can be detrimental to read speeds. A compressed chunk should be >1MB, <10MB (??TODO verify) for optimal reads.
```
Need to escape the "<" character here to fix build error "Unexpected character `1` (U+0031) before name, expected a character that can start a name, such as a letter, `$`, or `_`".

```suggestion
2. Concatenating several data variables together **if** a single chunk size is too small (\<1MB), at the expense of reducing readability of the Zarr store. Having too many small chunks can be detrimental to read speeds. A compressed chunk should be >1MB, \<10MB (??TODO verify) for optimal reads.
```

Also, reminder that we need to verify if these chunk sizes (1MB, 10MB) are correct, my memory is a bit foggy now :face_in_clouds:

negin513 replied:
Thanks! I spent some time trying to figure what went wrong in my local deployment and then I say your comment. 😉

negin513 replied:
I changed this to be between 1MB and 100MB. 5-100 MB per chunk is along the line of what we recommend for lustre. Our chunk sizes were around 6.25 MiB.

weiji14 commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,223 @@
+
+Here are some screenshots of the profiling results:
+
+- ![image](https://hackmd.io/_uploads/H1N1TTtAyl.png)
```
@negin513, could you provide push permissions to your fork? I can help copy the images from hackmd to here.

negin513 replied:
@weiji14 done. ✅ 

Thanks @dcherian , @TomAugspurger , and @weiji14 for your comments and for reviewing this post.

weiji14 commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,223 @@
+- GPU Direct Storage (GDS) for optimal performance
+- NVIDIA DALI
+- Work out how to use GDS when reading from cloud object store instead of on-prem disk.
+- etc
```
Want to shout out that reading/writing Zarr shards with GPU buffers (thanks @maxrjones and @TomAugspurger!) at https://github.com/zarr-developers/zarr-python/pull/2978 was just merged, and could go in here or somewhere above, depending on when this blog post gets published.

negin513 replied:
Next, we should test sharding on Lustre with and w/o GPU buffers. 🙈

dcherian commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,245 @@
+    github: negin513
+  - name: Wei Ji Leong
+    github: weiji14
+  - name: Max Jones
+    github: maxjones
```
```suggestion
  - name: Max Jones
    github: maxrjones
```

maxrjones commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,333 @@
+  />
+</div>
+
+We further quantified this bottleneck by comparing data loading and training throughput, as shown in the figure below:
```
```suggestion
We further quantified this bottleneck by comparing data loading and training throughput, as shown in the figure below (higher bars/more throughput is better):
```

maxrjones commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,333 @@
+
+In the plot above, the three bars represent:
+
+- Baseline: Baseline throughput of the end-to-end pipeline using real data.
```
```suggestion
- Training (real data): Baseline throughput of the end-to-end pipeline using real data.
```

maxrjones commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,333 @@
+1. **Optimized Chunking & Compression**
+   - We explored different chunking and compression strategies to optimize the data loading performance. We found that using Zarr v3 with optimized chunking and compression significantly improved the data loading performance.
+2. **GPU-native data loading with Zarr v3 and KvikIO**
+   - Leveraging Zarr v3's support for reading data directly into GPU memory using CuPy arrays, we utilized [KvikIO](https://docs.rapids.ai/api/kvikio/stable/) to bypass CPU memory, enabling direct data transfer from storage to GPU.
```
```suggestion
   - Leveraging Zarr Python 3's support for reading data directly into GPU memory using CuPy arrays, we utilized [KvikIO](https://docs.rapids.ai/api/kvikio/stable/) to bypass CPU memory, enabling direct data transfer from storage to GPU.
```

maxrjones commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,333 @@
+
+1. **Optimized Chunking & Compression**
+   - We explored different chunking and compression strategies to optimize the data loading performance. We found that using Zarr v3 with optimized chunking and compression significantly improved the data loading performance.
+2. **GPU-native data loading with Zarr v3 and KvikIO**
```
```suggestion
2. **GPU-native data loading with Zarr Python 3 and KvikIO**
```

maxrjones commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,333 @@
+### Step 1: Optimized chunking & Compression
+
+The ERA-5 dataset we were using had a sub-optimal chunking scheme of `{'time': 10, 'channel': C, 'height': H, 'width': W}`, which meant that a minimum of 10 timesteps of data was being read even if we only needed 2 consecutive timesteps at a time.
+We decided to rechunk the data to align with our access pattern of 1-timestep at a time, while reformating to Zarr v3.
```
```suggestion
We decided to rechunk the data to align with our access pattern of 1-timestep at a time, while reformating to Zarr format 3.
```

maxrjones commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,333 @@
+
+### Step 2: Reading with Zarr-Python v3 + kvikIO 📖
+
+[Zarr v3](https://zarr.dev/blog/zarr-python-3-release/) introduces new capabilities, including the ability to [read from Zarr stores into CuPy arrays (i.e. GPU memory)](https://github.com/zarr-developers/zarr-python/issues/2574).
```
```suggestion
[Zarr Python 3](https://zarr.dev/blog/zarr-python-3-release/) introduces new capabilities, including the ability to [read from Zarr stores into CuPy arrays (i.e. GPU memory)](https://github.com/zarr-developers/zarr-python/issues/2574).
```

maxrjones commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,333 @@
+   - Alternatively, wait for [sharding](https://zarr.readthedocs.io/en/stable/user-guide/performance.html#sharding) to be supported for GPU buffers in Zarr-python.
+3. Align chunks with model access pattern.
+
+The plot below shows the performance of the original dataset vs. the rechunked dataset (to optimal chunk size) vs. uncompressed Zarr v3 dataset.
```
```suggestion
The plot below shows the performance of the original dataset vs. the rechunked dataset (to optimal chunk size) vs. uncompressed Zarr format 3 dataset.
```

maxrjones commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,333 @@
+> ## Lessons Learned 💡
+
+> - **Chunking matters!** It really does and can make a huge difference in performance.
+> - **Zarr v3 enables GPU-native workflows**: Zarr v3 introduces experimental support for reading data directly into GPU memory via `zarr.config.enable_gpu()`. However, this is currently limited to the final stage of the codec pipeline, with decompression still handled by the CPU. We are working on enabling GPU-native decompression using `nvComp` to eliminate the host-device transfer.
```
```suggestion
> - **Zarr Python 3 enables GPU-native workflows**: Zarr Python 3 introduces experimental support for reading data directly into GPU memory via `zarr.config.enable_gpu()`. However, this is currently limited to the final stage of the codec pipeline, with decompression still handled by the CPU. We are working on enabling GPU-native decompression using `nvComp` to eliminate the host-device transfer.
```

maxrjones commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,333 @@
+
+To benefit from these new features, we recommend installing:
+
+- [`zarr>=3.0.3`](https://github.com/zarr-developers/zarr-python/releases/tag/v3.0.3)
```
```suggestion
- [`zarr>=3.0.8`](https://github.com/zarr-developers/zarr-python/releases/tag/v3.0.8)
```

negin513 replied:
Perfect. Thanks @weiji14 & @maxrjones

maxrjones commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,333 @@
+
+Note that using `engine="zarr"` like above would still result in data being loaded into CPU memory before being transferred to GPU memory.
+
+If your system supports [GPU Direct Storage (GDS)](https://developer.nvidia.com/blog/gpudirect-storage/), you can use the `kvikio` to read data directly into GPU memory, bypassing CPU memory.
```
```suggestion
If your system supports [GPU Direct Storage (GDS)](https://developer.nvidia.com/blog/gpudirect-storage/), you can use `kvikio` to read data directly into GPU memory, bypassing CPU memory.
```

maxrjones commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,333 @@
+airt.to_zarr(store="/tmp/air-temp.zarr", mode="w", zarr_format=3, consolidated=False)
+
+with zarr.config.enable_gpu():
+    ds = xr.open_dataset("/tmp/air-temp.zarr", engine="zarr", consolidated=False)
```
Can you explain why not to use consolidated metadata for direct to GPU reading? I am surprised about this because I would expect it to improve performance.

maxrjones commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,333 @@
+
+![Flowchart-technically decompression is still done on CPUs](/posts/gpu-pipline/flowchart_2.png)
+
+Eventually with this [cupy-xarray Pull Request merged](https://github.com/xarray-contrib/cupy-xarray/pull/70) (based on earlier work at https://xarray.dev/blog/xarray-kvikio), this can be simplified to:
```
```suggestion
Eventually with this [cupy-xarray Pull Request merged](https://github.com/xarray-contrib/cupy-xarray/pull/70) (based on earlier work at [https://xarray.dev/blog/xarray-kvikio](https://xarray.dev/blog/xarray-kvikio)), this can be simplified to:
```

maxrjones commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,333 @@
+
+Figure above shows benchmark comparing CPU vs GPU-based decompression, with or without GDS enabled using [the data reading benchmark here](https://github.com/pangeo-data/ncar-hackathon-xarray-on-gpus/blob/main/benchmark/era5_zarr_benchmark.py).
+
+![GPU native decompression](/posts/gpu-pipline/zstd_benchmark.png)
```
do you know why it's slower with GDS?

negin513 replied:
Potentially GDS compat mode. I ran a new set of tests and updated the figure per our previous meeting.

maxrjones commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,333 @@
+
+Next, checkout the [end-to-end example](https://github.com/pangeo-data/ncar-hackathon-xarray-on-gpus/tree/main/zarr_ML_optimization) directory, where we show how to integrate the DALI pipeline into a PyTorch DataLoader and training loop. This example demonstrates how to use DALI to load data from Zarr stores, preprocess it on the GPU, and feed it into a PyTorch model for training.
+
+Profiling results from the DALI pipeline demonstrate effective overlap between CPU and GPU workloads, significantly reducing GPU idle time (blue) and increasing overall training throughput:
```
the GPU activity is blue, right? rather than the idle time

maxrjones commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,333 @@
+> - **Chunking matters!** It really does and can make a huge difference in performance.
+> - **Zarr v3 enables GPU-native workflows**: Zarr v3 introduces experimental support for reading data directly into GPU memory via `zarr.config.enable_gpu()`. However, this is currently limited to the final stage of the codec pipeline, with decompression still handled by the CPU. We are working on enabling GPU-native decompression using `nvComp` to eliminate the host-device transfer.
+> - **GPU Direct Storage (GDS)** can be an improvement for data-intensive workflows, but requires some setup and configuration.
+> - **Compression trade-offs**: Using compression can reduce the amount of data transferred, but can also increase the time spent on decompression. We found that using Zarr v3 with GPU-based decompression can significantly improve performance.
+> - **NVIDIA DALI** is a powerful tool for optimizing data loading, but requires some effort to integrate into existing workflows.
+> - **CuPy-Xarray integration** is still a work in progress, but can be very useful for GPU-native workflows. Please see this PR for more details: [xarray-contrib/cupy-xarray#70](https://github.com/xarray-contrib/cupy-xarray/pull/70).
+> - **GPU-native decompression** is a promising area for future work, but full support (e.g. GPU-side Zstd decompression) requires further development and testing.
```
```suggestion
> - **Compression trade-offs**: Using compression can reduce the amount of data transferred, but can also increase the time spent on decompression. We found that using Zarr v3 with GPU-based decompression can significantly improve performance.
> - **GPU-native decompression** is a promising area for future work, but full support (e.g. GPU-side Zstd decompression) requires further development and testing.
> - **NVIDIA DALI** is a powerful tool for optimizing data loading, but requires some effort to integrate into existing workflows.
> - **CuPy-Xarray integration** is still a work in progress, but can be very useful for GPU-native workflows. Please see this PR for more details: [xarray-contrib/cupy-xarray#70](https://github.com/xarray-contrib/cupy-xarray/pull/70).
> - **NVIDIA Nsight** provides a [powerful tool](https://developer.nvidia.com/nsight-systems) for identifying bottlenecks.
```

maxrjones reviewed (commented):
This is so great, thank you @negin513!!!

weiji14 commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,322 @@
+
+- GPU Direct Storage (GDS) for optimal performance
+- NVIDIA DALI
+- Support for sharded Zarr with GPU-friendly access patterns [already merged in Zarr v3]().
```
```suggestion
- Support for sharded Zarr with GPU-friendly access patterns already [merged](https://github.com/zarr-developers/zarr-python/pull/2978) in Zarr v3.0.8.
```

weiji14 commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,322 @@
+    github: weiji14
+  - name: Max Jones
+    github: maxrjones
+  - name: Akshay Subranian
```
```suggestion
  - name: Akshay Subramaniam
```

weiji14 commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,322 @@
+
+ML pipelines for large scientific datasets typically include steps:
+
+- Reading raw data from disk or object storage (often CPU-bound)
```
```suggestion
- Reading raw data from disk or object storage (often IO-bound)
```

negin513 replied:
Good catch!

weiji14 commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,322 @@
+- Transforming / preprocessing data (often CPU-bound)
+- Model Training/Inference (often GPU-bound)
+
+Although GPU compute is incredibly fast, the CPU can become a bottleneck when dealing with large datasets. In an ideal scenario, we want to saturate the GPU with data as quickly as possible to minimize idle time on both the CPU and GPU.
```
```suggestion
Although GPU compute is incredibly fast, IO and CPU bottlenecks can be a pain when dealing with large datasets. In an ideal scenario, we want to saturate the GPU with data as quickly as possible to minimize idle time on both the CPU and GPU.
```

weiji14 commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,322 @@
+This will read the data directly from the Zarr store to GPU memory, significantly reducing I/O latency, especially for large datasets.
+However, it relies on the [NVIDIA GPUDirect Storage (GDS)](https://docs.nvidia.com/datacenter/pgp/gds/index.html) feature to be enabled and correctly configured on your system.
+
+**Note**: Even with GDS, the decompression step is still occurs on the CPU (see next section for GPU solutions!). This means that the data is still being decompressed on the CPU before being transferred to the GPU. However, this is still a significant improvement over the previous method, as it reduces the amount of data that needs to be transferred over the PCIe bus. In the figure below, we show ...
```
```suggestion
**Note**: Even with GDS, the decompression step will still occur on the CPU (see next section for GPU solutions!). This means that the data is still being decompressed on the CPU before being transferred to the GPU. However, this is still a significant improvement over the previous method, as it reduces the amount of data that needs to be transferred over the PCIe bus. In the figure below, we show the flowchart of the data loading process with GDS enabled (i.e. using `kvikio`):
```

weiji14 commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,322 @@
+
+For a fully GPU-native pipline, the decompression step should also be done on the GPU. This is where [NVIDIA's nvCOMP](https://developer.nvidia.com/nvcomp) library comes in. nvCOMP provides fast, GPU-native implementations of popular compression algorithms like Zstandard (Zstd)
+
+With nvCOMP, all steps of data loading including reading, decompressing, and transforming data can be done on the GPU, significantly reducing the time spent on data loading. Here is a flowchart of the data loading process with GDS and GPU-based decompression enabled:
```
```suggestion
With nvCOMP, all steps of data loading including reading from disk, decompression, and transforming data can be done on the GPU, significantly reducing the time spent on data loading. Here is a flowchart of the data loading process with GDS and GPU-based decompression enabled:
```

weiji14 commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,322 @@
+
+> These results show that GPU-based decompression can significantly reduce the time spent on data loading and cut I/O latency from storage to device (less data transfer over PCIe/NVLink). This is especially useful for large datasets, as it allows for faster data loading and processing.
+
+Keep an eye on this space, as we are working on integrating this into the Zarr ecosystem to enable GPU-based decompression for Zarr stores. This will allow for a fully GPU-native workflow, where all steps of data loading, including reading, decompressing, and transforming data, can be done on the GPU.
```
```suggestion
Keep an eye on this space, as we are working on integrating this into the Zarr ecosystem to enable GPU-based decompression for Zarr stores. This will allow for a fully GPU-native workflow, where all steps of data loading, including reading, decompression, and transforming data, can be done on the GPU.
```

weiji14 reviewed (commented):
Awesome work, this is coming along really nicely already! Just some minor nitpicks, but hope that we can publish this next month!

kafitzgerald reviewed (commented):
Thanks so much for putting this together!  

Mostly just a few minor suggestions from my end beyond the existing comments / questions.

negin513 commented on src/posts/gpu-pipeline/index.md:
```diff
@@ -0,0 +1,322 @@
+
+Although GPU compute is incredibly fast, the CPU can become a bottleneck when dealing with large datasets. In an ideal scenario, we want to saturate the GPU with data as quickly as possible to minimize idle time on both the CPU and GPU.
+
+In this hackathon, we explored several strategies to reduce the data loading bottleneck and build a GPU-native pipeline to maximize GPU utilization.
```
Slightly reworded to address. ;-)

negin513 commented:
Hey folks, I think I have addressed all comments & added all the plots, if either one of you can review this: all in all +DALI +nvcomp ~17x improvement... 💪

dcherian commented:
merging in time for a scipy talk!
